src/pages/BackgroundSignIn.js

```diff
@@ -0,0 +1,141 @@
+
+export default function BackgroundSignIn(props) {
+  const location = useLocation();
+  const history = useHistory();
+  const queryParams = new URLSearchParams(location.search);
+  let token = queryParams.get("token");
+  let { onSignIn, onError, bearerToken } = props;
+  token = bearerToken || (token || '');
+  let [accessToken, setAccessToken] = useState(token);
+  let [formToken, setFormToken] = useState(token);
+  let [invalidToken, setInvalidToken] = useState(false);
+
+  useEffect(() => {
+    const getUser = async () => {
+      return await User.getMe();
+    };
+
+    const redirect = (user) => {
+      const page = user.isAdmin ? 'admin_console' :
+        user.isChairPerson ? 'chair_console' :
+          user.isMember ? 'member_console' :
+            user.isResearcher ? 'dataset_catalog' :
+              user.isDataOwner ? 'data_owner_console' :
+                user.isAlumni ? 'summary_votes' : '/';
+      history.push(page);
+      if (onSignIn)
+        onSignIn();
+    };
+
+    const setUserRoleStatuses = (user) => {
+      const currentUserRoles = user.roles.map(roles => roles.name);
+      user.isChairPerson = currentUserRoles.indexOf(USER_ROLES.chairperson) > -1;
+      user.isMember = currentUserRoles.indexOf(USER_ROLES.member) > -1;
+      user.isAdmin = currentUserRoles.indexOf(USER_ROLES.admin) > -1;
+      user.isResearcher = currentUserRoles.indexOf(USER_ROLES.researcher) > -1;
+      user.isDataOwner = currentUserRoles.indexOf(USER_ROLES.dataOwner) > -1;
+      user.isAlumni = currentUserRoles.indexOf(USER_ROLES.alumni) > -1;
+      Storage.setCurrentUser(user);
+      return user;
+    };
+
+    const setIsLogged = (user) => {
+      Storage.setUserIsLogged(true);
+    };
+
+    const performLogin = () => {
+      if (accessToken) {
+        Storage.setGoogleData({ accessToken: accessToken });
+        getUser().then(
+          user => {
+            user = Object.assign(user, setUserRoleStatuses(user));
+            setIsLogged();
+            redirect(user);
+          },
+          error => {
+            const status = error.status;
+            switch (status) {
+              case 400:
+                if (onError)
+                  onError(error);
+                break;
+              case 409:
+                // If the user exists, just log them in.
+                getUser().then(
+                  user => {
+                    user = Object.assign(user, setUserRoleStatuses(user));
+                    setIsLogged();
+                    redirect(user);
+                  },
+                  error => {
+                    Storage.clearStorage();
+                  });
+                break;
+              case 401:
+                setInvalidToken(true);
+                break;
+              default:
+                setInvalidToken(true);
+                break;
+            }
+          });
+      }
+    };
+
+    if (accessToken)
+      performLogin();
```

Comment:
You also have an if statement for accessToken within performLogin itself, so you can remove the if block either within the function from this highlighted context. I lean towards removing it from the function since it simplifies it's logic (albeit slightly), but it's a matter of preference.

Reply:
That is a good call. I think removing it from the function makes sense.